### DISCUSSION TITLE
What could cause "avr-g++: error: device-specs/specs-attiny402: No such file or directory" Error?

### DISCUSSION THREAD

**Discussion by Peaser** [SpenceKonde/megaTinyCore#837] [Q&A]
Error appears when trying to compile for attiny402. May not necessarily matter, but the code is:

`#define F_CPU 20000000
#include <avr/delay.h>
#include <avr/io.h>

void setup() {
  PORTA.DIRSET |= PIN6_bm;
}

void loop() {
  VPORTA.OUT |= PIN6_bm;
  _delay_ms(1000);
  VPORTA.OUT |= ~PIN6_bm;
  _delay_ms(1000);
}`

I'm using arduino 1.8.8. Would updating to 1.8.16 fix it?

**Comment by SpenceKonde** [comment]
Might be old version of Arduino. The "golden version" right now is 1.8.13. 1.8.12 and all earlier versions have serious bugs in the programmer selection menus. 
That was fixed in 1.8.13, to much rejoicing. 

That rejoicing was shortlived, as 1.8.14 and all released versions after that contain a regression that interferes with manual installation of some cores. 

Note that **if you installed the core manually and did not follow the instructions in Installation.md** for manually upgrading the toolchain, that is the expected result. Manual installation of this core will not work for any thus-far released version of Arduino for all parts without also updating the toolchain. Depending on the version of Arduino, some parts may work, and some parts may partially work. Hence, the core should not be installed manually unless you are willing to jump through the hoops described and update the toolchain too. 

**I would not be surprised** if issues like that could result if you do a manual installation of this core (even following the instructions mentioned above) and also do board manager installation of other packages with certain combinations of IDE version, core version, and cores installed with board manager and manually. I do not have any reports of this situation, but I do not think it is possible that there is no such combination that will break - there is a fundamental problem with the hardware package specification - a manually installed hardware package is underspecified: Nowhere does it declare what version of the toolchain it needs (much less where to get it). That's in the board manager json file. What you have installed via board manager can thus change what toolchain versions the IDE has available to it - it has to decide whether to use the version of the toolchain it came with, or one that board manager installed. It does not always make wise decisions here, and I've gotten complaints about one of my other cores causing breakage of some other core. The other core was last updated when dinosaurs walked the earth, and when any of my cores were installed through board manager, the toolchain got updated, and now the no-longer-supported avr-gcc-specific behavior that they relied on broke. 

For that reason, we advise that users not manually install megaTinyCore, DxCore, or ATTinyCore on a non-portable Arduino installation on a system which has board packages installed via board manager. 
Even more importantly, we strongly advise against installing any of these cores via board manager on a non-portable Arduino instance, if you currently have a manually installed core that has not been recently updated. 

**We recommend board manager installation for all users not performing development work on the core itself** as this requires just copy-pasting the json url into preferences, and then using the simple board manager interface (tools -> board -> board manager) to install, upgrade or downgrade the core. Since there will be  a 2.6.2 in not that long to pick up a fix for at least one stupid bug, you should expect that you will need to upgrade soon ;-)

For users who are working on the core, my recommendation would be a manual installation, with manual toolchain update in a board-manager-never-used environment - that is, if a non-portable Arduino instance is used with board manager, perform the manual installations in a portable one, and vise-versa.

**Reply by Peaser** [reply]
I installed via preferences > http://drazzy.com/package_drazzy.com_index.json  > Boards manager, so it's assumed that this would work?

I also am compiling for a jtag2updi (sketch uploaded to a Nano), could that be responsible? It's very vague.

Here's the full error dump:
https://pastebin.com/f99pkp5N

I read elsewhere that to fix this specific "specs not found" error, One should delete  "GCC EXEC PREFIX" from system variables, I surprisingly never even had it yet still get the bug.

**Comment by hmeijdam** [comment]
Your code compiles fine on a newer IDE version (1.8.18) with 2.6.1 MegaTinycore
So I recommend you do an additional [portable installation](https://www.arduino.cc/en/Guide/PortableIDE) of a newer IDE version (so you can keep your current IDE version should you need it).
